### ISSUE TITLE
MaaS r11.1.0rc5: `Verify presence of expected alarms` fails

### ISSUE THREAD

**Issue by byronmccollum** [rcbops/rpc-openstack#621]
Seems the alarms are not fully created in MaaS before the `Verify presence of expected alarms` task is run. You can see the filesystem alarm for `/srv/node/swift4` on `ee149569-object1` failed, but the other 4 alarms were present:

```
failed: [ee149569-object1] => (item=filesystem_.srv.node.swift4--ee149569-object1) => {"failed": true, "item": "filesystem_.srv.node.swift4--ee149569-object1"}
msg: Expected alarm for check (filesystem_.srv.node.swift4--ee149569-object1) not found
skipping: [ee149569-object1] => (item=filesystem_.srv.node.swift1--ee149569-object1)
skipping: [ee149569-object1] => (item=filesystem_.srv.node.swift2--ee149569-object1)
skipping: [ee149569-object1] => (item=filesystem_.srv.node.swift3--ee149569-object1)
skipping: [ee149569-object1] => (item=filesystem_.srv.node.swift5--ee149569-object1)
```

### Full Task Output
```
TASK: [Verify presence of expected alarms] ************************************ 
failed: [ee149569-object1] => (item=filesystem_.srv.node.swift4--ee149569-object1) => {"failed": true, "item": "filesystem_.srv.node.swift4--ee149569-object1"}
msg: Expected alarm for check (filesystem_.srv.node.swift4--ee149569-object1) not found
failed: [ee149569-object1] => (item=swift_account_server_check--ee149569-object1) => {"failed": true, "item": "swift_account_server_check--ee149569-object1"}
msg: Expected alarm for check (swift_account_server_check--ee149569-object1) not found
failed: [ee149569-object1] => (item=swift_quarantine_check--ee149569-object1) => {"failed": true, "item": "swift_quarantine_check--ee149569-object1"}
msg: Expected alarm for check (swift_quarantine_check--ee149569-object1) not found
skipping: [ee149569-infra1] => (item=conntrack_count--ee149569-infra1)
skipping: [ee149569-infra1] => (item=cpu_check--ee149569-infra1)
skipping: [ee149569-infra1] => (item=disk_utilisation--ee149569-infra1)
skipping: [ee149569-infra1] => (item=filesystem_.--ee149569-infra1)
skipping: [ee149569-infra1] => (item=galera_check--ee149569-infra1_galera_container-0679ccad)
skipping: [ee149569-infra1] => (item=keystone_api_local_check--ee149569-infra1_keystone_container-135bbf6a)
skipping: [ee149569-infra1] => (item=lb_api_check_keystone)
skipping: [ee149569-infra1] => (item=lb_api_check_swift_proxy)
skipping: [ee149569-infra1] => (item=memcached_status--ee149569-infra1_memcached_container-8cd1e7e1)
skipping: [ee149569-infra1] => (item=memory_check--ee149569-infra1)
skipping: [ee149569-infra1] => (item=rabbitmq_status--ee149569-infra1_rabbit_mq_container-91c0c43b)
skipping: [ee149569-infra1] => (item=swift_proxy_server_check--ee149569-infra1_swift_proxy_container-ec9ea791)
skipping: [ee149569-object1] => (item=conntrack_count--ee149569-object1)
skipping: [ee149569-object1] => (item=cpu_check--ee149569-object1)
skipping: [ee149569-object1] => (item=disk_utilisation--ee149569-object1)
skipping: [ee149569-object1] => (item=filesystem_.--ee149569-object1)
skipping: [ee149569-object1] => (item=filesystem_.srv.node.swift1--ee149569-object1)
skipping: [ee149569-object1] => (item=filesystem_.srv.node.swift2--ee149569-object1)
skipping: [ee149569-object1] => (item=filesystem_.srv.node.swift3--ee149569-object1)
skipping: [ee149569-object1] => (item=filesystem_.srv.node.swift5--ee149569-object1)
skipping: [ee149569-object1] => (item=memory_check--ee149569-object1)
skipping: [ee149569-object1] => (item=swift_account_replication_check--ee149569-object1)
skipping: [ee149569-object1] => (item=swift_async_check--ee149569-object1)
skipping: [ee149569-object1] => (item=swift_container_replication_check--ee149569-object1)
skipping: [ee149569-object1] => (item=swift_container_server_check--ee149569-object1)
skipping: [ee149569-object1] => (item=swift_md5_check--ee149569-object1)
skipping: [ee149569-object1] => (item=swift_object_replication_check--ee149569-object1)
skipping: [ee149569-object1] => (item=swift_object_server_check--ee149569-object1)
skipping: [ee149569-object2] => (item=conntrack_count--ee149569-object2)
skipping: [ee149569-object2] => (item=cpu_check--ee149569-object2)
skipping: [ee149569-object2] => (item=disk_utilisation--ee149569-object2)
skipping: [ee149569-object2] => (item=filesystem_.--ee149569-object2)
skipping: [ee149569-object2] => (item=filesystem_.srv.node.swift1--ee149569-object2)
skipping: [ee149569-object2] => (item=filesystem_.srv.node.swift2--ee149569-object2)
skipping: [ee149569-object2] => (item=filesystem_.srv.node.swift3--ee149569-object2)
skipping: [ee149569-object2] => (item=filesystem_.srv.node.swift4--ee149569-object2)
skipping: [ee149569-object2] => (item=filesystem_.srv.node.swift5--ee149569-object2)
skipping: [ee149569-object2] => (item=memory_check--ee149569-object2)
skipping: [ee149569-object2] => (item=swift_account_replication_check--ee149569-object2)
skipping: [ee149569-object2] => (item=swift_account_server_check--ee149569-object2)
skipping: [ee149569-object2] => (item=swift_async_check--ee149569-object2)
skipping: [ee149569-object2] => (item=swift_container_replication_check--ee149569-object2)
skipping: [ee149569-object2] => (item=swift_container_server_check--ee149569-object2)
skipping: [ee149569-object2] => (item=swift_md5_check--ee149569-object2)
skipping: [ee149569-object2] => (item=swift_object_replication_check--ee149569-object2)
skipping: [ee149569-object2] => (item=swift_object_server_check--ee149569-object2)
skipping: [ee149569-object2] => (item=swift_quarantine_check--ee149569-object2)
skipping: [ee149569-logger1] => (item=conntrack_count--ee149569-logger1)
skipping: [ee149569-logger1] => (item=cpu_check--ee149569-logger1)
skipping: [ee149569-logger1] => (item=disk_utilisation--ee149569-logger1)
skipping: [ee149569-logger1] => (item=filesystem_.--ee149569-logger1)
skipping: [ee149569-logger1] => (item=memory_check--ee149569-logger1)
skipping: [ee149569-object3] => (item=conntrack_count--ee149569-object3)
skipping: [ee149569-object3] => (item=cpu_check--ee149569-object3)
skipping: [ee149569-object3] => (item=disk_utilisation--ee149569-object3)
skipping: [ee149569-object3] => (item=filesystem_.--ee149569-object3)
skipping: [ee149569-object3] => (item=filesystem_.srv.node.swift1--ee149569-object3)
skipping: [ee149569-object3] => (item=filesystem_.srv.node.swift2--ee149569-object3)
skipping: [ee149569-object3] => (item=filesystem_.srv.node.swift3--ee149569-object3)
skipping: [ee149569-object3] => (item=filesystem_.srv.node.swift4--ee149569-object3)
skipping: [ee149569-object3] => (item=filesystem_.srv.node.swift5--ee149569-object3)
skipping: [ee149569-object3] => (item=memory_check--ee149569-object3)
skipping: [ee149569-object3] => (item=swift_account_replication_check--ee149569-object3)
skipping: [ee149569-object3] => (item=swift_account_server_check--ee149569-object3)
skipping: [ee149569-object3] => (item=swift_async_check--ee149569-object3)
skipping: [ee149569-object3] => (item=swift_container_replication_check--ee149569-object3)
skipping: [ee149569-object3] => (item=swift_container_server_check--ee149569-object3)
skipping: [ee149569-object3] => (item=swift_md5_check--ee149569-object3)
skipping: [ee149569-object3] => (item=swift_object_replication_check--ee149569-object3)
skipping: [ee149569-object3] => (item=swift_object_server_check--ee149569-object3)
skipping: [ee149569-object3] => (item=swift_quarantine_check--ee149569-object3)
```

**Comment by byronmccollum** [comment]
Previous was a Standalone Swift deployment...Seeing the same issue on a Full RPC deployment, but obviously different alarms failing their presence checks:

```
TASK: [Verify presence of expected alarms] ************************************ 
skipping: [f82bebd6-block1] => (item=cinder_volume_check--f82bebd6-block1)
skipping: [f82bebd6-block1] => (item=conntrack_count--f82bebd6-block1)
skipping: [f82bebd6-block1] => (item=cpu_check--f82bebd6-block1)
skipping: [f82bebd6-block1] => (item=disk_utilisation--f82bebd6-block1)
skipping: [f82bebd6-block1] => (item=filesystem_.--f82bebd6-block1)
skipping: [f82bebd6-block1] => (item=memory_check--f82bebd6-block1)
skipping: [f82bebd6-logger1] => (item=conntrack_count--f82bebd6-logger1)
skipping: [f82bebd6-logger1] => (item=cpu_check--f82bebd6-logger1)
skipping: [f82bebd6-logger1] => (item=disk_utilisation--f82bebd6-logger1)
skipping: [f82bebd6-logger1] => (item=filesystem_.--f82bebd6-logger1)
skipping: [f82bebd6-logger1] => (item=memory_check--f82bebd6-logger1)
skipping: [f82bebd6-compute1] => (item=conntrack_count--f82bebd6-compute1)
skipping: [f82bebd6-compute1] => (item=cpu_check--f82bebd6-compute1)
skipping: [f82bebd6-compute1] => (item=disk_utilisation--f82bebd6-compute1)
skipping: [f82bebd6-compute1] => (item=filesystem_.--f82bebd6-compute1)
skipping: [f82bebd6-compute1] => (item=memory_check--f82bebd6-compute1)
skipping: [f82bebd6-compute1] => (item=neutron_linuxbridge_agent_check--f82bebd6-compute1)
skipping: [f82bebd6-compute1] => (item=nova_compute_check--f82bebd6-compute1)
skipping: [f82bebd6-object1] => (item=conntrack_count--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=cpu_check--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=disk_utilisation--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=filesystem_.--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=filesystem_.srv.node.swift1--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=filesystem_.srv.node.swift2--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=filesystem_.srv.node.swift3--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=filesystem_.srv.node.swift4--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=filesystem_.srv.node.swift5--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=memory_check--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=swift_account_replication_check--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=swift_account_server_check--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=swift_async_check--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=swift_container_replication_check--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=swift_container_server_check--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=swift_md5_check--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=swift_object_replication_check--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=swift_object_server_check--f82bebd6-object1)
skipping: [f82bebd6-object1] => (item=swift_quarantine_check--f82bebd6-object1)
failed: [f82bebd6-infra1] => (item=cinder_api_local_check--f82bebd6-infra1_cinder_api_container-86b71333) => {"failed": true, "item": "cinder_api_local_check--f82bebd6-infra1_cinder_api_container-86b71333"}
msg: Expected alarm for check (cinder_api_local_check--f82bebd6-infra1_cinder_api_container-86b71333) not found
failed: [f82bebd6-infra1] => (item=cinder_scheduler_check--f82bebd6-infra1_cinder_scheduler_container-f34d2118) => {"failed": true, "item": "cinder_scheduler_check--f82bebd6-infra1_cinder_scheduler_container-f34d2118"}
msg: Expected alarm for check (cinder_scheduler_check--f82bebd6-infra1_cinder_scheduler_container-f34d2118) not found
skipping: [f82bebd6-infra1] => (item=conntrack_count--f82bebd6-infra1)
failed: [f82bebd6-infra1] => (item=cpu_check--f82bebd6-infra1) => {"failed": true, "item": "cpu_check--f82bebd6-infra1"}
msg: Expected alarm for check (cpu_check--f82bebd6-infra1) not found
skipping: [f82bebd6-infra1] => (item=disk_utilisation--f82bebd6-infra1)
skipping: [f82bebd6-infra1] => (item=filesystem_.--f82bebd6-infra1)
failed: [f82bebd6-infra1] => (item=galera_check--f82bebd6-infra1_galera_container-08d97c89) => {"failed": true, "item": "galera_check--f82bebd6-infra1_galera_container-08d97c89"}
msg: Expected alarm for check (galera_check--f82bebd6-infra1_galera_container-08d97c89) not found
failed: [f82bebd6-infra1] => (item=galera_check--f82bebd6-infra1_galera_container-6652a395) => {"failed": true, "item": "galera_check--f82bebd6-infra1_galera_container-6652a395"}
msg: Expected alarm for check (galera_check--f82bebd6-infra1_galera_container-6652a395) not found
failed: [f82bebd6-infra1] => (item=galera_check--f82bebd6-infra1_galera_container-b0710277) => {"failed": true, "item": "galera_check--f82bebd6-infra1_galera_container-b0710277"}
msg: Expected alarm for check (galera_check--f82bebd6-infra1_galera_container-b0710277) not found
skipping: [f82bebd6-infra1] => (item=glance_api_local_check--f82bebd6-infra1_glance_container-79e2eed5)
skipping: [f82bebd6-infra1] => (item=glance_registry_local_check--f82bebd6-infra1_glance_container-79e2eed5)
skipping: [f82bebd6-infra1] => (item=heat_api_local_check--f82bebd6-infra1_heat_apis_container-897ba019)
failed: [f82bebd6-infra1] => (item=heat_cfn_api_check--f82bebd6-infra1_heat_apis_container-897ba019) => {"failed": true, "item": "heat_cfn_api_check--f82bebd6-infra1_heat_apis_container-897ba019"}
msg: Expected alarm for check (heat_cfn_api_check--f82bebd6-infra1_heat_apis_container-897ba019) not found
failed: [f82bebd6-infra1] => (item=heat_cw_api_check--f82bebd6-infra1_heat_apis_container-897ba019) => {"failed": true, "item": "heat_cw_api_check--f82bebd6-infra1_heat_apis_container-897ba019"}
msg: Expected alarm for check (heat_cw_api_check--f82bebd6-infra1_heat_apis_container-897ba019) not found
skipping: [f82bebd6-infra1] => (item=horizon_local_check--f82bebd6-infra1_horizon_container-69095b09)
skipping: [f82bebd6-infra1] => (item=keystone_api_local_check--f82bebd6-infra1_keystone_container-338abbd7)
failed: [f82bebd6-infra1] => (item=lb_api_check_cinder) => {"failed": true, "item": "lb_api_check_cinder"}
msg: Expected alarm for check (lb_api_check_cinder) not found
failed: [f82bebd6-infra1] => (item=lb_api_check_glance) => {"failed": true, "item": "lb_api_check_glance"}
msg: Expected alarm for check (lb_api_check_glance) not found
failed: [f82bebd6-infra1] => (item=lb_api_check_heat_api) => {"failed": true, "item": "lb_api_check_heat_api"}
msg: Expected alarm for check (lb_api_check_heat_api) not found
failed: [f82bebd6-infra1] => (item=lb_api_check_heat_cfn) => {"failed": true, "item": "lb_api_check_heat_cfn"}
msg: Expected alarm for check (lb_api_check_heat_cfn) not found
failed: [f82bebd6-infra1] => (item=lb_api_check_heat_cloudwatch) => {"failed": true, "item": "lb_api_check_heat_cloudwatch"}
msg: Expected alarm for check (lb_api_check_heat_cloudwatch) not found
failed: [f82bebd6-infra1] => (item=lb_api_check_horizon) => {"failed": true, "item": "lb_api_check_horizon"}
msg: Expected alarm for check (lb_api_check_horizon) not found
skipping: [f82bebd6-infra1] => (item=lb_api_check_keystone)
skipping: [f82bebd6-infra1] => (item=lb_api_check_neutron)
skipping: [f82bebd6-infra1] => (item=lb_api_check_nova)
skipping: [f82bebd6-infra1] => (item=lb_api_check_swift_proxy)
skipping: [f82bebd6-infra1] => (item=memcached_status--f82bebd6-infra1_memcached_container-d0d06873)
skipping: [f82bebd6-infra1] => (item=memory_check--f82bebd6-infra1)
failed: [f82bebd6-infra1] => (item=neutron_api_local_check--f82bebd6-infra1_neutron_server_container-3d0a5a31) => {"failed": true, "item": "neutron_api_local_check--f82bebd6-infra1_neutron_server_container-3d0a5a31"}
msg: Expected alarm for check (neutron_api_local_check--f82bebd6-infra1_neutron_server_container-3d0a5a31) not found
skipping: [f82bebd6-infra1] => (item=neutron_dhcp_agent_check--f82bebd6-infra1_neutron_agents_container-f41797f5)
skipping: [f82bebd6-infra1] => (item=neutron_l3_agent_check--f82bebd6-infra1_neutron_agents_container-f41797f5)
failed: [f82bebd6-infra1] => (item=neutron_linuxbridge_agent_check--f82bebd6-infra1_neutron_agents_container-f41797f5) => {"failed": true, "item": "neutron_linuxbridge_agent_check--f82bebd6-infra1_neutron_agents_container-f41797f5"}
msg: Expected alarm for check (neutron_linuxbridge_agent_check--f82bebd6-infra1_neutron_agents_container-f41797f5) not found
skipping: [f82bebd6-infra1] => (item=neutron_metadata_agent_check--f82bebd6-infra1_neutron_agents_container-f41797f5)
skipping: [f82bebd6-infra1] => (item=neutron_metadata_agent_proxy_check--f82bebd6-infra1_neutron_agents_container-f41797f5)
skipping: [f82bebd6-infra1] => (item=neutron_metering_agent_check--f82bebd6-infra1_neutron_agents_container-f41797f5)
skipping: [f82bebd6-infra1] => (item=nova_api_local_check--f82bebd6-infra1_nova_api_os_compute_container-f6f4ef5d)
skipping: [f82bebd6-infra1] => (item=nova_api_metadata_local_check--f82bebd6-infra1_nova_api_metadata_container-44d788d8)
skipping: [f82bebd6-infra1] => (item=nova_cert_check--f82bebd6-infra1_nova_cert_container-e2350add)
skipping: [f82bebd6-infra1] => (item=nova_conductor_check--f82bebd6-infra1_nova_conductor_container-68c3c8c1)
skipping: [f82bebd6-infra1] => (item=nova_consoleauth_check--f82bebd6-infra1_nova_console_container-3f2d7ef4)
skipping: [f82bebd6-infra1] => (item=nova_scheduler_check--f82bebd6-infra1_nova_scheduler_container-31453cd7)
skipping: [f82bebd6-infra1] => (item=nova_spice_console_check--f82bebd6-infra1_nova_console_container-3f2d7ef4)
skipping: [f82bebd6-infra1] => (item=rabbitmq_status--f82bebd6-infra1_rabbit_mq_container-b1bae977)
skipping: [f82bebd6-infra1] => (item=rabbitmq_status--f82bebd6-infra1_rabbit_mq_container-d8e89812)
skipping: [f82bebd6-infra1] => (item=rabbitmq_status--f82bebd6-infra1_rabbit_mq_container-e95e96be)
skipping: [f82bebd6-infra1] => (item=swift_proxy_server_check--f82bebd6-infra1_swift_proxy_container-e6c6c68b)
```

**Comment by andymcc** [comment]
@byronmccollum I'm battling to recreate this - is there something special I can do to get this to fail out?
I have a full install, but it successfully ran through the check once MaaS alarms/checks were setup - and i didn't get any failures.

Its interesting that it worked for the object2/object3 servers - and only not for object1.

Those checks do look like they should exist - I'm wondering if this task actually did what it meant to do (e.g. alerted to the fact that alarms/checks weren't there properly) - and that the actual maas setup failed in some way.

**Comment by git-harry** [comment]
It would help to look at the agent logs for the node that failed. Was it a race and the alarm was setup after the verify playbook was started or did the agent fail? Did you try running the verify-maas.yml playbook manually?

**Comment by sigmavirus24** [comment]
Hey @byronmccollum did you try re-running a few minutes later?

@stevelle @robb-romans can we link to the docs that discuss the need to wait a bit?

**Comment by d34dh0r53** [comment]
Closing this as it is a documented issue that has a reasonable workaround.

**Comment by byronmccollum** [comment]
Where is the workaround documented? Is using `deploy.sh` still the expected deployment mechanism? There is a 30 second sleep in there between `setup-maas.yml` and `verify-maas.yml`. Is there a better way of handling this besides an arbitrary sleep value? Seems like this is going to crop up again in the future...if not every single time...

**Comment by sigmavirus24** [comment]
So, the original issue did not mention this was coming from the deploy.sh.

@andymcc @stevelle I'm not sure `verify-maas.yml` should be in there, do you two agree?

**Comment by sigmavirus24** [comment]
@byronmccollum so, the thing is that `verify-maas.yml` needs a longer sleep in there for larger environments. The `rackspace-monitoring-agent` registers the checks for us with MaaS but it takes a quite a while and it seems to take longer the larger the environment is. We have it documented that you'll have to retry that but if after a reasonable amount of time it's still failing, you will have to investigate manually what checks are missing and why.

**Comment by byronmccollum** [comment]
Can a retries / until / delay be added in the playbooks instead of the arbitrary sleep that won't always work...?

**Comment by byronmccollum** [comment]
...or should `verify-maas.yml` even be a part of `deploy.sh` as you were inquiring...?

**Comment by stevelle** [comment]
No number of retries / delays can be chosen to always work, either. we could increase the sleep/delay and/or add retries but some day it will fail on a big deploy anyway.

There was an ask to have the verify be part of each run but I agree it is a problematic requirement. I would not be opposed to removing it from deploy.sh and adding an echo msg reminding the user to do it manually.

I think the doc around this was built into upgrades only, so that is also something to note. We were not aware of this happening during deploy.sh in the original report so saying it was a doc'd issue was incorrect and reopening this for further work might make sense.

**Comment by git-harry** [comment]
Ansible doesn't have a way to retry multiple tasks so I think this has to live in the bash script or be run manually.

**Comment by hughsaunders** [comment]
This should be fixed in #719 by moving the logic into rpc-maas-tool and retrying the task that runs that.

**Comment by robb-romans** [comment]
@hughsaunders thanks
